### PR TITLE
Add Syslog ILogger provider

### PR DESCRIPTION
**Pull request by oblaise**
## Description
This PR add support for a Logging provider that sends logs using Syslog protocol

## Motivation and Context
Syslog is a common logging provider with server present on most Un*x distributions. Servers also exist for Windows.

## How Has This Been Tested?
Used in a project. Logs are sent to a Syslog server present in an opensource project: https://github.com/PJO2/tftpd64/
It is hard to create unit test as you need a server on the other side.

## Types of changes
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: nanoFramework.Logging.Syslog/Facility.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/Facility.cs:1]: You are missing the headers here. Please copy/paste from the existing files

File: nanoFramework.Logging.Syslog/Severity.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/Severity.cs:1]: missing headers here as well

File: nanoFramework.Logging.Syslog/SyslogClient.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
+﻿using System;
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogClient.cs:1]: And here :-) I guess most likely in other places. Please add them in all the cs file

File: nanoFramework.Logging.Syslog/SyslogClient.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        /// </summary>
+        /// <param name="endpoint">Endpoint of the server</param>
+        /// <param name="localHostname">'Hostname' part of the RFC3164 message</param>
+        /// <param name="facility">Facility used by this logger</param>
+        /// <param name="localAddress">Local IP address to bind socket (if null IPAdress.Any will be used)</param>
+        /// <param name="localPort">Local port to bind socket (0 to choose available port)</param>
+        public SyslogClient(IPEndPoint endpoint, string localHostname, Facility facility = default, IPAddress localAddress = null, int localPort = 0)
+        {
+            if (localHostname is null)
+                throw new ArgumentNullException(nameof(localHostname));
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogClient.cs:32]: We are using brakets in all the code for the if statements. now, there is a better way to handle this one, you can replace this if statement and the LoclHostname line by:
```csharp
LocalHostname = localHostname ?? throw new ArgumentNullException(nameof(localHostname));
```

File: nanoFramework.Logging.Syslog/SyslogClient.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        /// </summary>
+        /// <param name="facility">Facility in the message</param>
+        /// <param name="severity">Severity in the message</param>
+        /// <param name="hostname">Hostname in the message</param>
+        /// <param name="messageTag">Message tag (will be followed by ': ' in the message)</param>
+        /// <param name="messageContent">Message content</param>
+        /// <exception cref="ObjectDisposedException"></exception>
+        public void SendMessage(Facility facility,Severity severity, string hostname, string messageTag, string messageContent)
+        {
+            if (_disposed)
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogClient.cs:86]: Please use the brackets { and }
Also please leave an extra enter avec the } So this code will then be:
```csharp
if (_disposed)
{
    throw new ObjectDisposedException(nameof(SyslogClient));
}

// We don't send message with Severity.None which isn't a RFC6134 valid value
if (severity != Severity.None)
{
     _socket.Send(FormatUdpMessage(facility, severity, hostname, messageTag, messageContent));
}
```

File: nanoFramework.Logging.Syslog/SyslogClient.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+            var headerBuilder = new StringBuilder();
+            // RFC3164 PRI
+            headerBuilder.Append('<').Append(priorityValue).Append('>');
+            // RFC3164 HEADER
+            string timestamp = DateTime.UtcNow.ToString("MMM dd HH:mm:ss");
+            headerBuilder.Append(timestamp).Append(' ');
+            headerBuilder.Append(localHostname).Append(' ');
+            // RFC3164 MSG
+            if (messageTag != null)
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogClient.cs:104]: brackets { }

File: nanoFramework.Logging.Syslog/SyslogClient.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                headerBuilder.Append(messageTag).Append(':').Append(' ');
+            headerBuilder.Append(messageContent ?? "");
+            var s = headerBuilder.ToString();
+            return UTF8Encoding.UTF8.GetBytes(s); // Ideally an UTF7 is required here to be RFC3164 compliant
+        }
+
+        private static IPEndPoint ResolveEndPoint(string hostname, int port)
+        {
+            IPHostEntry hostEntry = Dns.GetHostEntry(hostname);
+            if (hostEntry is null)
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogClient.cs:114]: brackets

File: nanoFramework.Logging.Syslog/SyslogClient.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                throw new ArgumentException($"Hostname `{hostname}` is not resolvable");
+            return new IPEndPoint(hostEntry.AddressList[0], port);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            _disposed = true;
+            ((IDisposable)_socket).Dispose();
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogClient.cs:123]: Just in case
```suggestion
            ((IDisposable)_socket)?.Dispose();
```

File: nanoFramework.Logging.Syslog/SyslogLogger.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace nanoFramework.Logging.Syslog
+{
+    /// <summary>
+    /// ILogger implmentation that send message log to a Syslog server following RFC3164 convention.
+    /// While not mandatory it is recommended to use <see cref="SyslogLoggerFactory"/> to create the SyslogLogger instances.
+    /// </summary>
+    public class SyslogLogger : ILogger
+    {
+        private SyslogClient _client;
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogLogger.cs:13]:
```suggestion
        private readonly SyslogClient _client;
```

File: nanoFramework.Logging.Syslog/SyslogLogger.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        /// </summary>
+        /// <param name="client">SyslogClient to use for sending message</param>
+        /// <param name="categoryName">CategoryName of this logger used as TAG in the Syslog messages</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        public SyslogLogger(SyslogClient client, string categoryName)
+        {
+            if (client is null)
+                throw new ArgumentNullException(nameof(client));
+            
+            _client = client;
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogLogger.cs:26]: remove the previous if and replace by:
```suggestion
            _client = client ?? throw new ArgumentNullException(nameof(client));
```

Comment (Ellerbach) [reply nanoFramework.Logging.Syslog/SyslogLogger.cs:26]: If there are other places, I'll just make a short version

File: nanoFramework.Logging.Syslog/SyslogLogger.cs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        /// <inheritdoc />
+        public bool IsEnabled(LogLevel logLevel) => logLevel >= MinLogLevel;
+
+        /// <inheritdoc />
+        public void Log(LogLevel logLevel, EventId eventId, string state, Exception exception, MethodInfo format)
+        {
+            if (logLevel >= MinLogLevel && logLevel!=LogLevel.None)
+            {
+                string message;
+                if (format is null)
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogLogger.cs:54]: brackets

File: nanoFramework.Logging.Syslog/SyslogLoggerFactory.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        /// <param name="localPort">Local port to bind socket (0 to choose available port)</param>
+        public SyslogLoggerFactory(string hostname, int port= SyslogClient.DefaultPort, string localHostname=DefaultHostname, Facility facility = default, IPAddress localAddress = null, int localPort = 0) 
+        {
+            _client = new SyslogClient(hostname, port, localHostname, facility , localAddress, localPort);
+        }
+
+        /// <inheritdoc/>
+        public ILogger CreateLogger(string categoryName)
+        {
+            if (_disposed)
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogLoggerFactory.cs:47]: brakets

File: nanoFramework.Logging.Syslog/SyslogLoggerFactory.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+                throw new ObjectDisposedException(nameof(SyslogLoggerFactory));
+            return new SyslogLogger(_client,categoryName);
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            _disposed = true;
+            // If logger created through this factory are used after this you will get exception
+            _client.Dispose();
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogLoggerFactory.cs:57]:
```suggestion
            _client?.Dispose();
```

**Review by Ellerbach — COMMENTED** [review]
Nice all up! Some adjustments needed in the code, missing headers as well, all up, real good!

File: nanoFramework.Logging.Syslog/Facility.cs [hunk 14]
```diff
@@ -0,0 +1,41 @@
+﻿//
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+using System;
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/Facility.cs:5]: You'll call me picky... But it does require an extra enter after the // and before the using.

File: nanoFramework.Logging.Syslog/SyslogLogger.cs [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                {
+                    message = $"{(string)format.Invoke(null, new object[] { LoggerCategoryName, logLevel, eventId, state, exception })}";
+                }
+
+                _client.SendMessage(LogLevelToSeverity(logLevel), LoggerCategoryName, message);
+            }
+        }
+        
+        private Severity LogLevelToSeverity(LogLevel level) => level switch
+            {
```
Comment (Ellerbach) [inline nanoFramework.Logging.Syslog/SyslogLogger.cs:69]: There is an indentation issue here. Use `ctrl + k + d`with Visual Studio, it will automatically format everything nicely.
And there is an extra enter after the function

**Review by Ellerbach — APPROVED** [review]
Looks good to me. Thanks again. If you can adjust the few 'picky' things, it would be perfect.

**Comment by Ellerbach** [comment]
@josesimoes the check for the versions does not understand the $version$, I guess those cases with $something$ should be excluded or just displayed as warning.

**Comment by oblaise** [comment]
> Looks good to me. Thanks again. If you can adjust the few 'picky' things, it would be perfect.

Done ;-)

**Comment by josesimoes** [comment]
> @josesimoes the check for the versions does not understand the $version$, I guess those cases with $something$ should be excluded or just displayed as warning.

@Ellerbach it does. The thing is that, despite being a valid nuspec value, the nuspec parser doesn't deal properly with these situations. A matter of adding this new assembly to the edge case list on nanovc.

**Comment by Ellerbach** [comment]
> A matter of adding this new assembly to the edge case list on nanovc.

Done :-) In PR :-) Will be a good test to see if all is in order now ;-)

**Comment by oblaise** [comment]
I saw a difference in the project file between syslog project and other projects. Crossing fingers.

**Comment by oblaise** [comment]
> I saw a difference in the project file between syslog project and other projects. Crossing fingers.

I see I'm missing the version of the packages in the nfproj project also. How can I add them ?
```
    <Reference Include="System.Threading">
      <HintPath>..\packages\nanoFramework.System.Threading.1.0.4-preview.8\lib\System.Threading.dll</HintPath>
    </Reference>
```

**Comment by josesimoes** [comment]
> > I saw a difference in the project file between syslog project and other projects. Crossing fingers.
> 
> I see I'm missing the version of the packages in the nfproj project also. How can I add them ?
> 
> ```
>     <Reference Include="System.Threading">
>       <HintPath>..\packages\nanoFramework.System.Threading.1.0.4-preview.8\lib\System.Threading.dll</HintPath>
>     </Reference>
> ```

@oblaise no need to worry with that. But if you want to add these, just copy from one of the other nfprojs that has them.

**Comment by oblaise** [comment]
architecture

> > > I saw a difference in the project file between syslog project and other projects. Crossing fingers.
> > 
> > 
> > I see I'm missing the version of the packages in the nfproj project also. How can I add them ?
> > ```
> >     <Reference Include="System.Threading">
> >       <HintPath>..\packages\nanoFramework.System.Threading.1.0.4-preview.8\lib\System.Threading.dll</HintPath>
> >     </Reference>
> > ```
> 
> @oblaise no need to worry with that. But if you want to add these, just copy from one of the other nfprojs that has them.

So I don't see what I can do more to have the version test to pass.

**Comment by josesimoes** [comment]
@oblaise please do not worry about the error in the nanovc check.
As Laurent already mentioned above that's something that needs to be adjusted in our tool. Not your fault.
It's been taken care of. Hopefully all will be OK in the next hour or so.

**Comment by josesimoes** [comment]
@oblaise build is 🟢 !!

**Comment by josesimoes** [comment]
I've fixed the package.config for you and pushed it to your working branch.

File: nanoFramework.Logging.Syslog/Facility.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+//
+
+using System;
+
+namespace nanoFramework.Logging.Syslog
+{
+	/// <summary>
+	/// Available facility types (RFC 3164).
+	/// </summary>
+	public enum Facility
```
Comment (josesimoes) [inline nanoFramework.Logging.Syslog/Facility.cs:13]: Wouldn't it be beneficial to have Intellisense comments here?
Or maybe just on the one(s) that this library supports?

On a second thought it could be even better to have here only the supported ones...

Comments?

Comment (oblaise) [reply nanoFramework.Logging.Syslog/Facility.cs:13]: `Facility` is an arbitrary value. Any of them can be used and is normally defined in the call to `SysLoggerFactory` constructor. For `Severity` effectively the mapping from `LogLevel` to Severity limits it to some of them. But You can always use `SysLogger.Client.SendMessage` methods if you want to use the severities that don't directly map to a `LogLevel` value.

**Comment by josesimoes** [comment]
@oblaise did one last pass with a few more improvements. Please review and comment back if you find anything wrong and/or misplaced.

**Comment by oblaise** [comment]
> OK for me. Give me a few minutes I will insert the comments in Facility/Severity.

@josesimoes how can I push your latest commit to my repo ?

**Review by josesimoes — APPROVED** [review]
All good now!!
